packages/connect-web/src/index.ts

```diff
@@ -1,107 +1,40 @@
-import { ConnectFactoryDependencies, factory } from '@trezor/connect/src/factory';
+import { factory } from '@trezor/connect/src/factory';
+import { TrezorConnectDynamic } from '@trezor/connect/src/impl/dynamic';
 import { CoreInIframe } from './impl/core-in-iframe';
 import { CoreInPopup } from './impl/core-in-popup';
-import ProxyEventEmitter from './utils/proxy-event-emitter';
-import type { ConnectSettings, ConnectSettingsPublic, Manifest } from '@trezor/connect/src/types';
-import EventEmitter from 'events';
-import { CallMethodPayload } from '@trezor/connect/src/events';
-import { getEnv } from './connectSettings';
-
-type TrezorConnectType = 'core-in-popup' | 'iframe';
+import type { ConnectSettingsPublic } from '@trezor/connect';
 
 const IFRAME_ERRORS = ['Init_IframeBlocked', 'Init_IframeTimeout', 'Transport_Missing'];
 
-/**
- * Implementation of TrezorConnect that can dynamically switch between iframe and core-in-popup implementations
- */
-export class TrezorConnectDynamicImpl implements ConnectFactoryDependencies {
-    public eventEmitter: EventEmitter;
-
-    private currentTarget: TrezorConnectType = 'iframe';
-    private coreInIframeImpl: CoreInIframe;
-    private coreInPopupImpl: CoreInPopup;
-
-    private lastSettings?: Partial<ConnectSettings>;
-
-    public constructor() {
-        this.coreInIframeImpl = new CoreInIframe();
-        this.coreInPopupImpl = new CoreInPopup();
-        this.eventEmitter = new ProxyEventEmitter([
-            this.coreInIframeImpl.eventEmitter,
-            this.coreInPopupImpl.eventEmitter,
-        ]);
-    }
-
-    private getTarget() {
-        return this.currentTarget === 'iframe' ? this.coreInIframeImpl : this.coreInPopupImpl;
-    }
-
-    private async switchTarget(target: TrezorConnectType) {
-        if (this.currentTarget === target) {
-            return;
+const impl = new TrezorConnectDynamic<'core-in-popup' | 'iframe'>({
+    implementations: [
+        {
+            type: 'iframe',
+            impl: new CoreInIframe(),
+        },
+        {
+            type: 'core-in-popup',
+            impl: new CoreInPopup(),
+        },
+    ],
+    getInitTarget: (settings: Partial<ConnectSettingsPublic>) => {
+        if (settings.coreMode === 'popup') {
+            return 'core-in-popup';
         }
 
-        await this.getTarget().dispose();
-        this.currentTarget = target;
-        await this.getTarget().init(this.lastSettings);
-    }
-
-    public manifest(manifest: Manifest) {
-        this.lastSettings = {
-            ...this.lastSettings,
-            manifest,
-        };
-
-        this.getTarget().manifest(manifest);
-    }
-
-    public async init(settings: Partial<ConnectSettingsPublic> = {}) {
-        const env = getEnv();
-        if (settings.coreMode === 'iframe' || settings.popup === false || env === 'webextension') {
```

Comment:
This condition now disappears but maybe it's not a problem. 
Before this meant that for `settings.popup === false || env === 'webextension'` it would not be in auto mode by default, it would strictly use iframe

Reply:
What happens if you have settings.coreMode === 'popup' in env === 'webextension'  ?

Is it possible to user coreMode === 'popup' in a webextension using connect-web?

Reply:
Undefined behavior... it could probably work, but I haven't tested it and I don't think there is any reason to use it.

Reply:
Right, and even webextension using connect-web is kind of legacy since it is not supported in MV3.
Ok, let's keep the previous logic c806de2c446d09781dee2bbd19dfe656e09459d5

Reply:
Well that condition should probably be put into `isCoreModeAuto`, because right now it doesn't affect anything in `getInitTarget`

Reply:
Do you mean in case when there is an error it should still keep 'iframe' in case that `env === 'webextenstion`?

Like:

```js
        const isCoreModeAuto =
            env !== 'webextension' ||
            impl.lastSettings?.coreMode === 'auto' ||
            impl.lastSettings?.coreMode === undefined;
```

Reply:
I just pushed a fixup that should be equivalent to how it worked previously

Reply:
:+1:  looks good to me, rebasing the PR...